shiny/ui/_theme.py

```diff
@@ -0,0 +1,364 @@
+from typing import Optional, TypeVar
+
+from htmltools import HTMLDependency
+from packaging.version import Version
+
+from .._docstring import no_example
+from .._versions import bootstrap
+from ._theme_presets import (
+    ShinyThemePreset,
+    shiny_theme_presets,
+    shiny_theme_presets_bundled,
+)
+from ._utils import path_pkg_www
+
+T = TypeVar("T", bound="Theme")
+
+
+@no_example()
+class Theme:
+    """
+    Create a custom Shiny theme.
+
+    The `Theme` class allows you to create a custom Shiny theme by providing custom Sass
+    code. The theme can be based on one of the available presets, such as `"shiny"` or
+    `"bootstrap"`, or a Bootswatch theme. Use the `.add_*()` methods can be chained
+    together to add custom Sass functions, defaults, mixins, and rules.
+
+    Pass the `Theme` object directly to the `theme` argument of any Shiny page function,
+    such as :func:`~shiny.ui.page_sidebar` or :func:`~shiny.ui.page_navbar`. In Shiny
+    Express apps, use the `theme` argument of :func:`~shiny.express.ui.page_opts` to set
+    the app theme.
+
+    Customized themes require the [libsass package](https://pypi.org/project/libsass/),
+    and are compiled to CSS when the theme is used. The `Theme` class caches the
+    compiled CSS, but you can speed up app loading (and avoid the runtime `libsass`
+    dependency) by pre-compiling the theme CSS and saving it to a file. To do this, use
+    the `.to_css()` method to render the theme to a single minified CSS string. Once
+    saved to a file, the CSS can be used in any Shiny app by passing the file path to
+    the `theme` argument described above.
+
+    Parameters
+    ----------
+    preset
+        The name of the preset to use as a base. `"shiny"` is the default theme for
+        Shiny apps and `"bootstrap"` uses standard Bootstrap 5 styling. Bootswatch theme
+        presets are also available. Use `Theme.available_presets()` to see the full
+        list.
+    name
+        A custom name for the theme. If not provided, the preset name will be used.
+
+    Raises
+    ------
+    ValueError
+        If the `preset` is not a valid theme preset.
+    """
+
+    def __init__(
+        self,
+        preset: ShinyThemePreset = "shiny",
+        name: Optional[str] = None,
+    ):
+        check_is_valid_preset(preset)
+        self._preset: ShinyThemePreset = preset
+        self.name = name
+        self._version = bootstrap
+
+        # User-provided Sass code
+        self._functions: list[str] = []
+        self._defaults: list[str] = []
+        self._mixins: list[str] = []
+        self._rules: list[str] = []
+
+        # _css is either:
+        # 1. "precompiled" indicating it's okay to use precompiled preset.min.css
+        # 2. "" indicating that the CSS has not been compiled yet
+        # 3. A string containing the compiled CSS for the current theme
+        self._css: str = "precompiled" if preset in shiny_theme_presets_bundled else ""
```

Comment:
I'm pretty sure I'd prefer a boolean `_css_precompiled` flag instead of overloading the meaning of `_css`

Reply:
This was refactored, `self._css` is either the CSS or an empty string now.

We now consult `self._can_use_precompiled()` to see if we can use the precompiled CSS, if so we read it into `self._css`.